<a href="https://colab.research.google.com/github/jproctor-rebecca/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/LS_DS_121_Statistics_Probability_Assignment_RJProctor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [1]:
### YOUR WORK HERE

# grab the file from UCI
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data



--2020-07-13 19:37:11--  https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25936 (25K) [application/x-httpd-php]
Saving to: ‘imports-85.data’

imports-85.data     100%[===================>]  25.33K  --.-KB/s    in 0.1s    

2020-07-13 19:37:11 (194 KB/s) - ‘imports-85.data’ saved [25936/25936]



In [8]:
import pandas as pd

column_headers = ['symboling', 'normalized-losses', 'make', 'fuel-type', 
                  'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 
                  'engine-location', 'wheel-base', 'length', 'width', 'height', 
                  'curb-weight', 'engine-type', 'num-of-cylinders', 
                  'engine-size', 'fuel-system', 'bore', 'stroke', 
                  'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 
                  'highway-mpg', 'price']
               
# add column headers and replace '?' with NaN values
cars = pd.read_csv('imports-85.data', names=column_headers, na_values="?")


cars.head()


symboling  normalized-losses         make  ... city-mpg highway-mpg    price
0          3                NaN  alfa-romero  ...       21          27  13495.0
1          3                NaN  alfa-romero  ...       21          27  16500.0
2          1                NaN  alfa-romero  ...       19          26  16500.0
3          2              164.0         audi  ...       24          30  13950.0
4          2              164.0         audi  ...       18          22  17450.0

[5 rows x 26 columns]

In [5]:
cars.describe()

symboling  normalized-losses  ...  highway-mpg         price
count  205.000000         164.000000  ...   205.000000    201.000000
mean     0.834146         122.000000  ...    30.751220  13207.129353
std      1.245307          35.442168  ...     6.886443   7947.066342
min     -2.000000          65.000000  ...    16.000000   5118.000000
25%      0.000000          94.000000  ...    25.000000   7775.000000
50%      1.000000         115.000000  ...    30.000000  10295.000000
75%      2.000000         150.000000  ...    34.000000  16500.000000
max      3.000000         256.000000  ...    54.000000  45400.000000

[8 rows x 16 columns]

In [6]:
cars.isnull().sum()

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

In [10]:
print(cars.shape)

(205, 26)


### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [33]:
### YOUR WORK HERE
import numpy as np
import scipy
from scipy import stats

sample = cars.sample(10, random_state=30)

sample

symboling  normalized-losses  ... highway-mpg    price
121          1              154.0  ...          38   6692.0
176         -1               65.0  ...          32  10898.0
24           1              148.0  ...          38   6229.0
88          -1              137.0  ...          30   9279.0
72           3              142.0  ...          18  35056.0
103          0              108.0  ...          25  13499.0
26           1              148.0  ...          38   7609.0
20           0               81.0  ...          43   6575.0
155          0               91.0  ...          32   8778.0
185          2               94.0  ...          34   8195.0

[10 rows x 26 columns]

In [34]:
sample.isnull().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                0
dtype: int64

In [12]:
sample['highway-mpg'].mean()

32.8

In [13]:
cars['highway-mpg'].mean()

30.75121951219512

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [14]:
### YOUR WORK HERE
sample['highway-mpg'].mean()


32.8

In [ ]:
# his claim cannot be verified though small batch sampling
# must apply statistical test for further analysis 

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [15]:
### YOUR WORK HERE
sample_100 = cars.sample(100, random_state=30)

sample_100.head()


symboling  normalized-losses  ... highway-mpg    price
121          1              154.0  ...          38   6692.0
176         -1               65.0  ...          32  10898.0
24           1              148.0  ...          38   6229.0
88          -1              137.0  ...          30   9279.0
72           3              142.0  ...          18  35056.0

[5 rows x 26 columns]

In [16]:
sample_100['highway-mpg'].mean()

30.64

In [ ]:
# it appears that as the sampling size increases the mean of 'highway-mpg' decreases; 
# the correlation proportion is not directly appearant & it is unknow if this is always true
# must apply statistical tests for further analysis

t-Test

1) Null Hypothesis:

Cars sold have average fuel efficiency of 35mpg on the highway: x¯¯¯==μ

2) Alternative Hypothesis:

Cars sold no NOT have average fuel efficiency of 35mpg on the highway: x¯¯¯≠μ

3) Confidence Level: 95%

In [57]:
sample_10 = cars.sample(10, random_state=30)
stats.ttest_1samp(sample_10, 0.09091)
### unsure of p-value = (0.045455+00.045455) ###
### what is ttest_1samp and where does it come from? ###
### when I read in the documentation, it looks like I should use my sample_10 here ###
### but in the lecture example Ryan pulled it out of thin air and did not define ###
### anywhere, but it still worked...this happens everyday & is why I stay confused ###
### about the code ###

sample_10.head()

### error here ###

TypeError: ignored

4) Conclusion:
Based on a t-statistic of 0.0, and a p-value of 0., (reject/***fail to reject***
) the null hypothesis that the cars sold have a mean fuel efficiency of 35mpg

t-Test

1) Null Hypothesis:

Cars sold have average fuel efficiency of 35mpg on the highway: x¯¯¯==μ

2) Alternative Hypothesis:

Cars sold no NOT have average fuel efficiency of 35mpg on the highway: x¯¯¯≠μ

3) Confidence Level: 95%

In [56]:
sample_100 = cars.sample(100, random_state=30)
stats.ttest_1samp(sample_100, 0.09091)
### unsure of p-value = (0.045455+00.045455) ###
### same issue as above ###

sample_100.head()

### error here ###

TypeError: ignored

4) Conclusion: Based on a t-statistic of 0.0, and a p-value of ~0, (***reject***
/fail to reject) the null hypothesis that the cars sold have a mean fuel efficiency of 35mpg.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

As the sample size increased, it became more reflective of the whole population, including outliers


## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [38]:
### YOUR WORK HERE
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data


--2020-07-13 22:58:31--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-07-13 22:58:32 (137 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [50]:
import numpy as np
import scipy
from scipy import stats

In [39]:
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

votes = pd.read_csv('house-votes-84.data', names=column_headers, na_values="?")

votes.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n          NaN
2    democrat                 NaN             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [40]:
votes_new = votes.replace({'y': 1, 'n': 0})

votes_new.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [42]:
votes_new.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           423.000000     387.000000  ...  407.000000    331.000000
mean              0.442080       0.503876  ...    0.427518      0.812689
std               0.497222       0.500632  ...    0.495327      0.390752
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      1.000000
50%               0.000000       1.000000  ...    0.000000      1.000000
75%               1.000000       1.000000  ...    1.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [43]:
votes_new.isnull().sum()

party                     0
handicapped-infants      12
water-project            48
budget                   11
physician-fee-freeze     11
el-salvador-aid          15
religious-groups         11
anti-satellite-ban       14
aid-to-contras           15
mx-missile               22
immigration               7
synfuels                 21
education                31
right-to-sue             25
crime                    17
duty-free                28
south-africa            104
dtype: int64

In [44]:
rep = votes_new[votes_new['party']=='republican']
dem = votes_new[votes_new['party']=='democrat']

In [45]:
print(rep.shape)
rep.head()

(168, 17)


party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [47]:
rep.isnull().sum()

party                    0
handicapped-infants      3
water-project           20
budget                   4
physician-fee-freeze     3
el-salvador-aid          3
religious-groups         2
anti-satellite-ban       6
aid-to-contras          11
mx-missile               3
immigration              3
synfuels                 9
education               13
right-to-sue            10
crime                    7
duty-free               12
south-africa            22
dtype: int64

In [46]:
print(dem.shape)
dem.head()

(267, 17)


party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [48]:
dem.isnull().sum()

party                    0
handicapped-infants      9
water-project           28
budget                   7
physician-fee-freeze     8
el-salvador-aid         12
religious-groups         9
anti-satellite-ban       8
aid-to-contras           4
mx-missile              19
immigration              4
synfuels                12
education               18
right-to-sue            15
crime                   10
duty-free               16
south-africa            82
dtype: int64

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

t-Test #1

1) Null Hypothesis:

Democrats support this bill at the same level that republicans do.

H0:x¯¯¯r==x¯¯¯d

2) Alternative Hypothesis:

Democrats support this bill in a different proportion than republicans.

Ha:x¯¯¯r≠x¯¯¯d

3) Confidence Level: 99%

In [54]:
stats.ttest_ind(rep['water-project'], dem['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

4) Conclusion:

Due to a t-statistic of 0.08 and a p-value of 0.92 we fail to reject the null hypothesis that democrats and repulicans support the water project bill at the same rate, and suggest the alternative that there are differing levels of support for this bill between the two parties.

t-Test #2

1) Null Hypothesis:

Democrats support this bill at the same level that republicans do.

H0:x¯¯¯r==x¯¯¯d

2) Alternative Hypothesis:

Democrats support this bill in a different proportion than republicans.

Ha:x¯¯¯r≠x¯¯¯d

3) Confidence Level: 99%1) Null Hypothesis:

Cars sold have average fuel efficiency of 35mpg on the highway: x¯¯¯==μ

2) Alternative Hypothesis:

Cars sold no NOT have average fuel efficiency of 35mpg on the highway: x¯¯¯≠μ

3) Confidence Level: 95%

In [55]:
stats.ttest_ind(rep['aid-to-contras'], dem['aid-to-contras'], nan_policy='omit')

Ttest_indResult(statistic=-18.052093200819733, pvalue=2.82471841372357e-54)

Due to a t-statistic of -18.08 and a p-value of 2.82 we fail to reject the null hypothesis that democrats and repulicans support the aid to contras bill at the same rate, and suggest the alternative that there are differing levels of support for this bill between the two parties.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE




### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE